File: tests/algorithms/x/Java/searches/median_of_medians.java
public class Main {

    static int[] set_at_int(int[] xs, int idx, int value) {
        int i = 0;
        int[] res = ((int[])(new int[]{}));
        while (i < xs.length) {
            if (i == idx) {
                res = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(res), java.util.stream.IntStream.of(value)).toArray()));
            } else {
                res = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(res), java.util.stream.IntStream.of(xs[i])).toArray()));
            }
            i = i + 1;
        }
        return res;
    }

    static int[] sort_int(int[] xs) {
        int[] res_1 = ((int[])(xs));
        int i_1 = 1;
        while (i_1 < res_1.length) {
            int key = res_1[i_1];
            int j = i_1 - 1;
            while (j >= 0 && res_1[j] > key) {
                res_1 = ((int[])(set_at_int(((int[])(res_1)), j + 1, res_1[j])));
                j = j - 1;
            }
            res_1 = ((int[])(set_at_int(((int[])(res_1)), j + 1, key)));
            i_1 = i_1 + 1;
        }
        return res_1;
    }

    static int median_of_five(int[] arr) {
        int[] sorted = ((int[])(sort_int(((int[])(arr)))));
        return sorted[Math.floorDiv(sorted.length, 2)];
    }

    static int median_of_medians(int[] arr) {
        if (arr.length <= 5) {
            return median_of_five(((int[])(arr)));
        }
        int[] medians = ((int[])(new int[]{}));
        int i_2 = 0;
        while (i_2 < arr.length) {
            if (i_2 + 5 <= arr.length) {
                medians = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(medians), java.util.stream.IntStream.of(median_of_five(((int[])(java.util.Arrays.copyOfRange(arr, i_2, i_2 + 5)))))).toArray()));
            } else {
                medians = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(medians), java.util.stream.IntStream.of(median_of_five(((int[])(java.util.Arrays.copyOfRange(arr, i_2, arr.length)))))).toArray()));
            }
            i_2 = i_2 + 5;
        }
        return median_of_medians(((int[])(medians)));
    }

    static int quick_select(int[] arr, int target) {
        if (target > arr.length) {
            return -1;
        }
        int x = median_of_medians(((int[])(arr)));
        int[] left = ((int[])(new int[]{}));
        int[] right = ((int[])(new int[]{}));
        boolean check = false;
        int i_3 = 0;
        while (i_3 < arr.length) {
            if (arr[i_3] < x) {
                left = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(left), java.util.stream.IntStream.of(arr[i_3])).toArray()));
            } else             if (arr[i_3] > x) {
                right = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(right), java.util.stream.IntStream.of(arr[i_3])).toArray()));
            } else             if (arr[i_3] == x) {
                if (!check) {
                    check = true;
                } else {
                    right = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(right), java.util.stream.IntStream.of(arr[i_3])).toArray()));
                }
            } else {
                right = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(right), java.util.stream.IntStream.of(arr[i_3])).toArray()));
            }
            i_3 = i_3 + 1;
        }
        int rank_x = left.length + 1;
        int answer = 0;
        if (rank_x == target) {
            answer = x;
        } else         if (rank_x > target) {
            answer = quick_select(((int[])(left)), target);
        } else {
            answer = quick_select(((int[])(right)), target - rank_x);
        }
        return answer;
    }

    static void main() {
        System.out.println(_p(median_of_five(((int[])(new int[]{5, 4, 3, 2})))));
        System.out.println(_p(quick_select(((int[])(new int[]{2, 4, 5, 7, 899, 54, 32})), 5)));
    }
    public static void main(String[] args) {
        main();
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/Kotlin/searches/median_of_medians.kt
import java.math.BigInteger

fun <T> _sliceList(lst: MutableList<T>, start: Int, end: Int): MutableList<T> {
    val st = if (start < 0) 0 else start
    val en = if (end > lst.size) lst.size else end
    if (st >= en) return mutableListOf()
    return lst.subList(st, en).toMutableList()
}

fun set_at_int(xs: MutableList<Int>, idx: Int, value: Int): MutableList<Int> {
    var i: Int = (0).toInt()
    var res: MutableList<Int> = mutableListOf<Int>()
    while (i < xs.size) {
        if (i == idx) {
            res = run { val _tmp = res.toMutableList(); _tmp.add(value); _tmp }
        } else {
            res = run { val _tmp = res.toMutableList(); _tmp.add(xs[i]!!); _tmp }
        }
        i = i + 1
    }
    return res
}

fun sort_int(xs: MutableList<Int>): MutableList<Int> {
    var res: MutableList<Int> = xs
    var i: Int = (1).toInt()
    while (i < res.size) {
        var key: Int = (res[i]!!).toInt()
        var j: BigInteger = ((i - 1).toBigInteger())
        while ((j.compareTo((0).toBigInteger()) >= 0) && (res[(j).toInt()]!! > key)) {
            res = set_at_int(res, ((j.add((1).toBigInteger())).toInt()), res[(j).toInt()]!!)
            j = j.subtract((1).toBigInteger())
        }
        res = set_at_int(res, ((j.add((1).toBigInteger())).toInt()), key)
        i = i + 1
    }
    return res
}

fun median_of_five(arr: MutableList<Int>): Int {
    var sorted: MutableList<Int> = sort_int(arr)
    return sorted[sorted.size / 2]!!
}

fun median_of_medians(arr: MutableList<Int>): Int {
    if (arr.size <= 5) {
        return median_of_five(arr)
    }
    var medians: MutableList<Int> = mutableListOf<Int>()
    var i: Int = (0).toInt()
    while (i < arr.size) {
        if ((i + 5) <= arr.size) {
            medians = run { val _tmp = medians.toMutableList(); _tmp.add(median_of_five(_sliceList(arr, i, i + 5))); _tmp }
        } else {
            medians = run { val _tmp = medians.toMutableList(); _tmp.add(median_of_five(_sliceList(arr, i, arr.size))); _tmp }
        }
        i = i + 5
    }
    return median_of_medians(medians)
}

fun quick_select(arr: MutableList<Int>, target: Int): Int {
    if (target > arr.size) {
        return 0 - 1
    }
    var x: Int = (median_of_medians(arr)).toInt()
    var left: MutableList<Int> = mutableListOf<Int>()
    var right: MutableList<Int> = mutableListOf<Int>()
    var check: Boolean = false
    var i: Int = (0).toInt()
    while (i < arr.size) {
        if (arr[i]!! < x) {
            left = run { val _tmp = left.toMutableList(); _tmp.add(arr[i]!!); _tmp }
        } else {
            if (arr[i]!! > x) {
                right = run { val _tmp = right.toMutableList(); _tmp.add(arr[i]!!); _tmp }
            } else {
                if (arr[i]!! == x) {
                    if (!check) {
                        check = true
                    } else {
                        right = run { val _tmp = right.toMutableList(); _tmp.add(arr[i]!!); _tmp }
                    }
                } else {
                    right = run { val _tmp = right.toMutableList(); _tmp.add(arr[i]!!); _tmp }
                }
            }
        }
        i = i + 1
    }
    var rank_x: Int = (left.size + 1).toInt()
    var answer: Int = (0).toInt()
    if (rank_x == target) {
        answer = x
    } else {
        if (rank_x > target) {
            answer = quick_select(left, target)
        } else {
            answer = quick_select(right, target - rank_x)
        }
    }
    return answer
}

fun user_main(): Unit {
    println(median_of_five(mutableListOf(5, 4, 3, 2)).toString())
    println(quick_select(mutableListOf(2, 4, 5, 7, 899, 54, 32), 5).toString())
}

fun main() {
    user_main()
}
